server/src/main/java/org/elasticsearch/plugins/ExtensiblePlugin.java

```diff
@@ -31,4 +31,10 @@
      * Reload any SPI implementations from the given classloader.
      */
     default void reloadSPI(ClassLoader loader) {}
+
+    /**
+     * Be notified of an extension plugin, will be called once per plugin having extendedPlugins pointing to this plugin.
+     * @param plugin the extending plugin
+     */
+    default void extensionPlugin(Plugin plugin) {}
```

Comment:
A method allowing `ExtensiblePlugin`'s to receive information on extending plugins.

Reply:
This seems backwards to me. Plugin A extending plugin B should only know about B through SPI classes. As I understand it, the autscaling case here needs to implement an autoscaling SPI interface inside ML, which wants ML info. So it's not the autoscaling Plugin that is needed, but the ML plugin where the concrete implementation exists.

Reply:
That is also how this works. I think the naming here may be the cause of confusion. The `plugin` passed as parameter is the extending pluging, i.e., ML in this case. Whereas the `this` object is `Autoscaling`. This method would thus notify `Autoscaling` of the existence of a plugin (ML) extending the autoscaling plugin.

Reply:
> This method would thus notify Autoscaling of the existence of a plugin (ML)

This is what I don't like. Autoscaling shouldn't be passed the instance of the ML plugin. It should find out about implementations of it's SPI interface through s service loader, ie through META-INF/services. There shouldn't be a need for it to know about the Plugin instances that extend its SPI interfaces.

Reply:
ML need it's Plugin instance in order to share it's memory tracker object to the decider service it is going to create. I am not sure how this can be done without sharing either the Plugin instance OR allowing a more magic service hook-up? (without resorting to static fields).

We could limit what the `ExtensiblePlugin` can use from the plugin to be the declared part as illustrated [here](https://github.com/elastic/elasticsearch/compare/master...henningandersen:enhance_autoscaling_plugability_alpha_variant1?expand=1#diff-e09cb2427887296e2ddd98d06d7d85ecR39), protected through a [proxy](https://github.com/elastic/elasticsearch/compare/master...henningandersen:enhance_autoscaling_plugability_alpha_variant1?expand=1#diff-40d98239fea26b07396b5032b95e396fR546). This would mean that `Autoscaling` is only allowed to call the methods from the `AutoscalingPlugin` interface.

But I am not sure this is your concern. Can you elaborate on why you dislike the approach?

Additionally, I think reusing the original Plugin object has other benefits:

1. Easier to add `ExtensiblePlugin`s (i.e., like the `Autoscaling` plugin allows pluggable decider services).
2. The resulting extension point interfaces (`AutoscalingPlugin` in this PR) looks and feels like all the core extension points (`SearchPlugin`), which also use a single Plugin instance.

Reply:
> ML need it's Plugin instance

From what I saw, it looked like ML is not getting it's plugin instance, but autoscaling is getting ML's Plugin instance. The purpose of ExtensiblePlugin is to allow triggering using ServiceLoader once new implementations are available. I think we should stick to that narrow design, and use change how we support SPI to optionally pass the Plugin instance to the implementation.

Reply:
Are you suggesting that ML also becomes an `ExtensiblePlugin`? I would find that odd, since it is not Extensible, only autoscaling is.

Or that we pass down ML's Plugin instance together with it's class-loader when calling `reloadSPI`? That is essentially the same as now, in that reloadSPI has to happen after loading the plugin. I can certainly rename the `extensionPlugin` method and pass down the ML Plugin instance too, as in `reloadSPI(ClassLoader, Plugin)`. But it does not fundamentally change anything in that Autoscaling would be passed the ML plugin instance.

I am not sure I see how Autoscaling can use the class-loader/service-loader to create an ML object that knows about  the ML plugin without passing the ML plugin to Autoscaling?

I am not sure of where you want to go. I think I need you to be a bit more specific to allow me to take this forward.